conformance/manifests/base.yaml

```diff
@@ -0,0 +1,300 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: gateway-conformance-infra
+  labels:
+    gateway-conformance: infra
+---
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: gateway-same-namespace
```

Comment:
Having more specific names with some semantic meaning like "shared-infra-gateway" would be better, especially as more and more tests use this.

Reply:
I've added some comments at the top of this file describing the key resources, but unfortunately don't have great ideas for how to make the names of the Gateways more descriptive right now. At this point, the only real variation between them is where they accept routes to bind from:

- gateway-same-namespace (only supports route in same ns)
- gateway-all-namespaces (supports routes in all ns)
- gateway-backend-namespaces (supports routes in ns with backend label)

In the future when we add resources with different protocols, multiple listeners, etc, naming may become a bit more clear. Open to ideas though.

Reply:
I know it's more verbose, but does it make sense to repeat the YAML across test cases (i.e. make each case hermetic) rather than trying to deduplicate? I know it's kind of a pain, but sharing the setup can result in coupling between test cases and the creation of a complex "base" case that makes it harder to understand the relationship later on.

Reply:
Yeah it's going to be tough to find the right balance here. For any route to work, it needs the following:

- Namespace
- Gateway
- Service
- Pods

Spinning all of those up and down for every test would quickly become quite time consuming. Alternatively if we were to leave them running indefinitely, we'd need each test case to be sure that it wasn't overlapping with resources defined by other test cases. This would also result in rather verbose yaml definitions for each test case. A final argument in favor of this shared set of base resources is that it enables conformance tests to be meaningful for implementations that are not currently able to support dynamically provisioning Gateways (this generally means there's a 1:1 relationship between Gateway and controller).

With all that said, more isolated tests would enable us to run tests in parallel which could significantly speed up a test run when there were a large number of tests in play. It would also be easier to understand each individual test case than our current approach that requires an understanding of the base resources.

I think we'll inevitably want to move towards a test suite that has more isolated test definitions, but fortunately that should be a fairly straightforward transition from this current state. I think that the current proposal is an acceptable starting point given the constraints above, but that we should try to avoid expanding the base set of resources, and in the future add test cases that are not entirely dependent on them.